File: mylib/query.py
"""Query the database"""

import sqlite3

DB = "../transfer.db"


def connect():
    return sqlite3.connect(DB)


# Create a new record (Create)
def create_transfer(
    connection,
    url: str,
    player_id: int,
    federation: str,
    former_fed: str,
    transfer_date: str,
):
    cursor = connection.cursor()
    cursor.execute(
        """
        INSERT INTO transfer (url, player_id, federation, former_fed, transfer_date)
        VALUES (?, ?, ?, ?, ?)
    """,
        (url, player_id, federation, former_fed, transfer_date),
    )
    connection.commit()
    print("Record inserted successfully")


# Read the latest record and return its ID (Read latest)
def read_latest_transfer(connection):
    cursor = connection.cursor()
    cursor.execute(
        """
        SELECT id, url, player_id, federation, former_fed, transfer_date
        FROM transfer
        ORDER BY id DESC
        LIMIT 1
    """
    )
    row = cursor.fetchone()
    if row:
        id, url, player_id, federation, former_fed, transfer_date = row
        print(
            f"""Latest Record - ID: {id},
        URL: {url}, Player ID: {player_id},
        Federation: {federation},
        Former Federation: {former_fed},
        Transfer Date: {transfer_date}"""
        )
        return id
    else:
        print("No records found.")
        return None


# Update a specific record (Update)
def update_transfer(connection, id: int, new_url: str):
    cursor = connection.cursor()
    cursor.execute(
        """
        UPDATE transfer
        SET url = ?
        WHERE id = ?
    """,
        (new_url, id),
    )
    connection.commit()
    print("Record updated successfully")


# Delete a specific record (Delete)
def delete_transfer(connection, id: int):
    cursor = connection.cursor()
    cursor.execute("DELETE FROM transfer WHERE id = ?", (id,))
    connection.commit()
    print("Record deleted successfully")


# Execute CRUD operations in sequence
def query():
    connection = connect()

    read_latest_transfer(connection)
    # Create a new record (Create)
    create_transfer(connection, "google.com", 123, "USA", "THA", "2004-02-01")

    # Read the latest record and get its ID (Read latest)
    latest_id = read_latest_transfer(connection)
    if latest_id is not None:
        # Update the latest record (Update)
        update_transfer(connection, latest_id, "example.com")

        # Delete the latest record (Delete)
        delete_transfer(connection, latest_id)

        read_latest_transfer(connection)
    else:
        print("No records found to update or delete.")

    connection.close()
    print("CRUD operations completed successfully")

File: rust-cli/src/query.rs
use rusqlite::{params, Connection, Result};

// Create a new record (Create)
pub fn create_transfer(
    connection: &Connection,
    url: &str,
    player_id: i32,
    federation: &str,
    former_fed: &str,
    transfer_date: &str,
) -> Result<()> {
    connection.execute(
        "INSERT INTO transfer (url, player_id, federation, former_fed, transfer_date)
         VALUES (?, ?, ?, ?, ?)",
        params![url, player_id, federation, former_fed, transfer_date],
    )?;
    println!("Record inserted successfully");
    Ok(())
}

// Read the latest record and return its ID (Read latest)
pub fn read_latest_transfer(connection: &Connection) -> Result<Option<i32>> {
    let mut stmt = connection.prepare(
        "SELECT id, url, player_id, federation, former_fed, transfer_date
         FROM transfer
         ORDER BY id DESC
         LIMIT 1",
    )?;

    let mut rows = stmt.query([])?;

    if let Some(row) = rows.next()? {
        let id: i32 = row.get(0)?;
        let url: String = row.get(1)?;
        let player_id: i32 = row.get(2)?;
        let federation: String = row.get(3)?;
        let former_fed: String = row.get(4)?;
        let transfer_date: String = row.get(5)?;

        println!(
            "Latest Record - ID: {}, URL: {}, Player ID: {}, Federation: {}, Former Federation: {}, Transfer Date: {}",
            id, url, player_id, federation, former_fed, transfer_date
        );

        Ok(Some(id))
    } else {
        println!("No records found.");
        Ok(None)
    }
}

// Update a specific record (Update)
pub fn update_transfer(connection: &Connection, id: i32, new_url: &str) -> Result<()> {
    connection.execute(
        "UPDATE transfer
         SET url = ?1
         WHERE id = ?2",
        params![new_url, id],
    )?;
    println!("Record updated successfully");
    Ok(())
}

// Delete a specific record (Delete)
pub fn delete_transfer(connection: &Connection, id: i32) -> Result<()> {
    connection.execute(
        "DELETE FROM transfer
         WHERE id = ?1",
        params![id],
    )?;
    println!("Record deleted successfully");
    Ok(())
}
